### PR TITLE
RSDK-5105 enable hardware pwm on custom linux

### PR DESCRIPTION
**Pull request by oliviamiller**
all pins that have both gpio and pwm defined in their board defintion files will be hardware pwm only.

### CODE REVIEW

**Review by penguinland — APPROVED** [review]
LGTM! When you mark the ticket as Done, remember to put a note for the docs team that the customlinux boards should have a warning that many boards can't do both hardware PWM and GPIO on the same pin, and that should be tested carefully when you write a new board definition file.

...do we need to deal with https://github.com/viamrobotics/rdk/pull/3357 and suggest that everyone specify null-like PWM data on every pin? Hopefully the answer is "no," but that PR makes me doubt myself.

**Comment by oliviamiller** [comment]
> LGTM! When you mark the ticket as Done, remember to put a note for the docs team that the customlinux boards should have a warning that many boards can't do both hardware PWM and GPIO on the same pin, and that should be tested carefully when you write a new board definition file.
> 
> ...do we need to deal with #3357 and suggest that everyone specify null-like PWM data on every pin? Hopefully the answer is "no," but that PR makes me doubt myself.

yes, I'm pretty sure that if a user wants to use software pwm on a pin they will need to set the pwmID as -1 in their board definitions. I can prioritize fixing this, think it should be an easy fix.

File: components/board/customlinux/board.go [hunk 1]
```diff
@@ -87,13 +87,13 @@ func parseRawPinData(pinData []byte, filePath string) ([]genericlinux.PinDefinit
...
 
-		// Until we get hardware PWM working reliably on lots of boards (likely by being able to
-		// set the pin mode directly), we disable all hardware PWM on these boards.
-		// The range operator makes a copy of the value, so to change the original, we need to look
-		// it up in the map again.
-		// TODO(RSDK-5105): Undo this when we're ready to support hardware PWM.
-		parsedPinData.Pins[name].PwmChipSysfsDir = ""
-		parsedPinData.Pins[name].PwmID = -1
+		// Until we can reliably switch between gpio and pwm on lots of boards, pins that have
+		// hardware pwm enabled will be hardware pwm only. Disabling gpio functianality on these
```
Comment (penguinland) [inline components/board/customlinux/board.go:91]: type: functionality

File: components/board/customlinux/board.go [hunk 2]
```diff
@@ -87,13 +87,13 @@ func parseRawPinData(pinData []byte, filePath string) ([]genericlinux.PinDefinit
...
+		// pins.
+		if parsedPinData.Pins[name].PwmChipSysfsDir != "" {
```
Comment (penguinland) [inline components/board/customlinux/board.go:93]: I'd also check that the line number is non-negative. Some people might be lazy and disable a pin by leaving the directory as-is and setting the line to a negative value.

**Review by penguinland — APPROVED** [review]
beyond that, LGTM!

File: components/board/customlinux/board.go (hunk 1 above)
Comment (randhid) [reply components/board/customlinux/board.go:91]: having a typo in a typo comment is so meta.

File: components/board/customlinux/board.go [hunk 3]
```diff
@@ -87,13 +87,13 @@ func parseRawPinData(pinData []byte, filePath string) ([]genericlinux.PinDefinit
...
+			logging.Global().Warnf("pin %s can be used for PWM only", parsedPinData.Pins[name].Name)
```
Comment (randhid) [inline components/board/customlinux/board.go:94]: can we add a to-do ticket to refactor the generic Linux constructor so it looks more like other component constructors? That way we can pass in the rdk logger and not have to rely on a global logger.

**Review by randhid — APPROVED** [review]

**Pull request merged**